stablehlo/reference/Element.cpp

```diff
@@ -175,63 +175,52 @@ bool areApproximatelyEqual(APFloat f, APFloat g) {
+                        debugString(value.getBitWidth()).c_str()));
   type_ = type;
-  value_ = APInt(type.getIntOrFloatBitWidth(), value,
-                 /*isSigned=*/isSupportedSignedIntegerType(type));
+  value_ = value;
 }
 
-Element::Element(Type type, bool value) : type_(type), value_(value) {}
-
-Element::Element(Type type, APFloat value) : type_(type), value_(value) {}
-
-Element::Element(Type type, double value) {
-  if (isSupportedFloatType(type)) {
-    APFloat floatVal(value);
-    bool roundingErr;
-    floatVal.convert(type.cast<FloatType>().getFloatSemantics(),
-                     APFloat::rmNearestTiesToEven, &roundingErr);
-    type_ = type;
-    value_ = floatVal;
-  } else if (isSupportedComplexType(type)) {
-    APFloat real(value);
-    APFloat imag(0.0);
-    auto floatTy = type.cast<ComplexType>().getElementType().cast<FloatType>();
-    bool roundingErr;
-    real.convert(floatTy.getFloatSemantics(), APFloat::rmNearestTiesToEven,
-                 &roundingErr);
-    imag.convert(floatTy.getFloatSemantics(), APFloat::rmNearestTiesToEven,
-                 &roundingErr);
-    type_ = type;
-    value_ = std::make_pair(real, imag);
-  } else {
+Element::Element(Type type, bool value) {
+  if (!isSupportedBooleanType(type))
     report_fatal_error(invalidArgument("Unsupported element type: %s",
                                        debugString(type).c_str()));
-  }
+  type_ = type;
+  value_ = value;
 }
 
-Element::Element(Type type, std::complex<APFloat> value)
-    : type_(type), value_(std::make_pair(value.real(), value.imag())) {}
+Element::Element(Type type, APFloat value) {
+  if (!isSupportedFloatType(type))
+    report_fatal_error(invalidArgument("Unsupported element type: %s",
+                                       debugString(type).c_str()));
+  if (type.getIntOrFloatBitWidth() != value.bitcastToAPInt().getBitWidth())
+    report_fatal_error(invalidArgument(
+        "Bitwidth mismatch. Type: %s, Value: %s",
+        debugString(type.getIntOrFloatBitWidth()).c_str(),
+        debugString(value.bitcastToAPInt().getBitWidth()).c_str()));
+  type_ = type;
+  value_ = value;
+}
 
-Element::Element(Type type, std::complex<double> value) {
+Element::Element(Type type, std::complex<APFloat> value) {
   if (!isSupportedComplexType(type))
     report_fatal_error(invalidArgument("Unsupported element type: %s",
                                        debugString(type).c_str()));
-  APFloat real(value.real());
-  APFloat imag(value.imag());
-  auto floatTy = type.cast<ComplexType>().getElementType().cast<FloatType>();
-  bool roundingErr;
-  real.convert(floatTy.getFloatSemantics(), APFloat::rmNearestTiesToEven,
-               &roundingErr);
-  imag.convert(floatTy.getFloatSemantics(), APFloat::rmNearestTiesToEven,
-               &roundingErr);
+  if (type.cast<ComplexType>().getElementType().cast<FloatType>().getWidth() !=
+      value.real().bitcastToAPInt().getBitWidth())
```

Comment:
I'll be removing this check as this also requires the same semantics as the `type`. Additionally, since `Element convert(Type type, std::complex<APFloat> value)` also calls `Element convert(Type type, APFloat value)` (which now also requires same semantics), I think it's okay to skip it here. I don't think we would run into a case where we accidentally have different semantics as it would be caught by the difference in type, and converting APFloat has been the convention for some time.

Reply:
I'm not sure I follow why remove the check.

"I don't think we would run into a case where we accidentally have different semantics". I think this is fairly easy to overlook. E.g. `map` functions do stuff like `return Element(type, floatFn(floatEl))`, and I'm not sure that it will be immediately evident to users of `map` that `floatFn` must return an APFloat with the same semantics as its argument.

And it's not like the check is hard to implement? Something like `type.cast<FloatType>().getSemantics() == value.getSemantics()`, right?

Reply:
Looks like I got feedbacks mixed in my head as I equated removing the comments on bitwidths as removing the semantics too. Added back now :)